tss-esapi/examples/hmac.rs

```diff
@@ -0,0 +1,171 @@
+        .build()
+        .unwrap();
+
+    let primary = context
+        .execute_with_nullauth_session(|ctx| {
+            // Create the key under the "owner" hierarchy. Other hierarchies are platform
+            // which is for boot services, null which is ephemeral and resets after a reboot,
+            // and endorsement which allows key certification by the TPM manufacturer.
+            ctx.create_primary(Hierarchy::Owner, primary_pub, None, None, None, None)
+        })
+        .unwrap();
+
+    // Create the HMAC key. This key exists under the primary key in it's hierarchy
+    // and can only be used if the same primary key is recreated from the parameters
+    // defined above.
+
+    let object_attributes = ObjectAttributesBuilder::new()
+        .with_fixed_tpm(true)
+        .with_fixed_parent(true)
+        .with_st_clear(false)
+        .with_sensitive_data_origin(true)
+        .with_user_with_auth(true)
+        // The key is used only for signing.
+        .with_sign_encrypt(true)
+        .build()
+        .expect("Failed to build object attributes");
+
+    let key_pub = PublicBuilder::new()
+        // This key is a HMAC key
+        .with_public_algorithm(PublicAlgorithm::KeyedHash)
+        .with_name_hashing_algorithm(HashingAlgorithm::Sha256)
+        .with_object_attributes(object_attributes)
+        .with_keyed_hash_parameters(PublicKeyedHashParameters::new(
+            KeyedHashScheme::HMAC_SHA_256,
+        ))
+        .with_keyed_hash_unique_identifier(Digest::default())
+        .build()
+        .unwrap();
+
+    let (enc_private, public) = context
+        .execute_with_nullauth_session(|ctx| {
+            // Create the HMAC key given our primary key as it's parent. This returns the private
+            // and public portions of the key. It's *important* to note that the private component
+            // is *encrypted* by a key associated with the primary key. It is not plaintext or
+            // leaked in this step.
+            ctx.create(primary.key_handle, key_pub, None, None, None, None)
+                .map(|key| (key.out_private, key.out_public))
+        })
+        .unwrap();
+
+    // Once the key is created, we have it's parameters in the private and public values.
+    // We now need to load it into the tpm so that it can be used.
+    //
+    // The enc_private and public values can be serialised and persisted - that way they can
+    // be reloaded for future use.
+
+    let input_data = MaxBuffer::try_from("TPMs are cool.".as_bytes().to_vec())
+        .expect("Failed to create buffer for input data.");
+
+    let hmac1 = context
+        .execute_with_nullauth_session(|ctx| {
+            // Load the HMAC key into the tpm context.
+            let hmac_key = ctx
+                .load(primary.key_handle, enc_private.clone(), public.clone())
+                .unwrap();
+            // Perform the HMAC.
+            let r = ctx.hmac(
+                hmac_key.clone().into(),
```

Comment:
That ```hmac_key``` is a ```KeyHandle``` and it derives the ```Copy``` trait so it will be copied. Just remove the cloning of the handle and all should work.

Reply:
Already done?

Reply:
Ahhh I'd missed one.